README.rst
===========================================================================
pg_fabrep: Postgresql Streaming Replication with Fabric and Repmgr
===========================================================================

Introduction
============

pg_fabrep allow you to easy setup a **Postgresql** 9.x **Hot Standby Streaming Replication** Server With **Repmgr**.

Patches are welcome! Feel free to fork and contribute to this project on:

**bitbucket**: `bitbucket.org/DNX/pg_fabrep <https://bitbucket.org/DNX/pg_fabrep/>`_


**github**: `github.com/DNX/pg_fabrep <https://github.com/DNX/pg_fabrep>`_


Installation
============

There are a few different ways to install pg_fabrep:

Using pip
---------
If you have pip install available on your system, just type::

    pip install pg_fabrep

If you've already got an old version of pg_fabrep, and want to upgrade, use::

    pip install -U pg_fabrep

Installing from a directory
---------------------------
If you've obtained a copy of pg_fabrep using either Mercurial or a downloadable
archive, you'll need to install the copy you have system-wide. Try running::

    python setup.py develop

If that fails, you don't have ``setuptools`` or an equivalent installed;
either install them, or run::

    python setup.py install


How to use pg_fabrep?
=====================

If you have already installed pg_fabrep, you must proceed with the
configuration of your cluster.

Configuration
-------------

Copy the scheleton for your cluster settings from <pg_fabrep> path::

    $ cp <pg_fabrep>/example_fabfile.py <your_project>/fabfile.py

How to find your <pg_fabrep> path? Just run::

    $ python -c "import pg_fabrep; print(pg_fabrep.__path__)"

or directly:

    $ cp `python -c "import pg_fabrep; print(pg_fabrep.__path__[0])"`/example_fabfile.py <your_project>/fabfile.py

In order to start **setup** task you must change only two parameters,
**pgmaster_ip** and **pgslave_ip** inside you cluster configuration task in earlier created **fabfile.py**, all other parameters have default
values and if you need you can change them too.

Here is a list with all parameters you can change and a short explanation for each:

- **env.configured**, don't change this, used for testing purposes
- **env.postgres_version**, postgres version, can be: "9.0" or "9.1", default: "9.1"
- **env.cluster_name**, name of your cluster - only numbers, letters and underscores, default: main
- **env.cluster_port**, postgres cluster port, default: 5432
- **env.pgmaster_ip**, the IP of your master server
- **env.master_ssh_port**, ssh port on the master server, will be used by repmgr/rsync from the slave to sync the files from the master, default: 22
- **env.master_node_number**, master node number, default: 1
- **env.pgmaster_user_host**, user@host used by fabric to establish a ssh tunnel between the machine from where the "setup" is launched and the master server, default: "root@<env.pgmaster_ip>"
- **env.master_pgconf_path**, master pgconf path, default: "/etc/postgresql/<env.postgres_version>/<cluster_name>/"
- **env.master_pgdata_path**, master pgdata path, default: "/var/lib/postgresql/<env.postgres_version>/<env.cluster_name>/"
- **env.pgslave_ip**, the IP of your slave server
- **env.slave_node_number**, slave node number, default: 2
- **env.pgslave_user_host**, user@host used by fabric to establish a ssh tunnel between the machine from where the "setup" is launched and the slave server, default: "root@<env.pgslave_ip>"
- **env.slave_pgconf_path**, slave pgconf path, default: "<env.master_pgconf_path>"
- **env.slave_pgdata_path**, slave pgdata path, default: "<env.master_pgdata_path>"
- **env.sync_db**, sync database used by repmgr, will be created if not exists, default: "syncdb"
- **env.sync_user**, sync postgres user used by repmgr, will be created if not exists, default: "syncuser"
- **env.sync_pass**, sync postgres user password, default: "syncpass"
- **env.repmgr_deb**, repmgr_deb installer, default:"postgresql-repmgr-<env.postgres_version>_1.0.0.deb"
- **env.ask_confirmation**, always ask user for confirmation when run any tasks, default: True


Please pay attention to not have any tasks in your fabfile.py called:
"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""

* setup

* test_configuration

or

* print_configuration

because these names are reserved by pg_fabrep.

Test your configuration first!
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
pg_fabrep come with its own automatic configuration tests. Each time you run
**setup** task, configuration tests are called.
Anyway, you can manually run these tests for your project configuration::

    fab <cluster_task_name> test_configuration

If you run **test_configuration** manually, you'll observe some output about all your project settings.

Do you need an example?
~~~~~~~~~~~~~~~~~~~~~~~

Ok, let's assume you want to configure a cluster called "cluster5444".
So, what we know about it?
we know:

* the cluster is called **cluster123**

* the postgres version we want to be used is 9.1

* the ip of the master server is: **11.11.11.11**

* the ip of the slave server is: **22.22.22.22**

* the port to be used in your cluster is 5444

* on each server is installed ubuntu server

Ok, it's enough to setup the replication, let's do it!
Create a folder to place you cluster settings::

    $ mkdir ~/clusters/
    $ cd ~/clusters/

copy here the example_fabfile.py from <pg_fabrep>::

    # we found pg_fabrep installation path first
    $ python -c "import pg_fabrep; print(pg_fabrep.__path__)"
    $ cp <pg_fabrep>/example_fabfile.py fabfile.py

now, in our current folder we have a file called **fabfile.py**
which is going to be edited according with our needs.

#. Change task name::

    # from:
    def example_cluster():
    # to:
    def cluster123():

#. Change env.pgmaster_ip::

    # from:
    env.pgmaster_ip = ""
    # to:
    env.pgmaster_ip = "11.11.11.11"

#. Change env.pgslave_ip::

    # from:
    env.pgslave_ip = ""
    # to:
    env.pgslave_ip = "22.22.22.22"

not, let's test our configuration::

    fab cluster123 test_configuration

you must see a message::

    Configuration tests passed!


Setup your replication
----------------------

Assuming you've set your cluster details in the fabfile.py now you are ready to launch the setup::

    fab cluster123 setup

during this process you can see all the output of the commands launched on
the master and slave servers. At some point you may be asked for some
information as sudo user password.
At the end of this task you must view a message saying that the setup
successful ended.
Now you can go on with the real tests of the replication.

CHANGELOG.rst
Changelog
=============================================================

Version 0.0.4
-------------------------------------------------------------

* ready for Ubuntu 12.04 LTS

Version 0.0.4
-------------------------------------------------------------

* added master_ssh_port parameter
* updated README.rst

Version 0.0.3
-------------------------------------------------------------

* don't stop postgres on the slave
* added postgres_version in the .bash_profile PGDATA

Version 0.0.2
-------------------------------------------------------------

* updated the documentation
* declared namespace in __init__.py for setuptools 0.7

Version 0.0.1
-------------------------------------------------------------

* initial structure

pypi_stats.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
Calculates the total number of downloads that a particular PyPI package has
received across all versions tracked by PyPI
"""

from datetime import datetime
import locale
import sys
import xmlrpclib

locale.setlocale(locale.LC_ALL, '')


class PyPIDownloadAggregator(object):

    def __init__(self, package_name, include_hidden=True):
        self.package_name = package_name
        self.include_hidden = include_hidden
        self.proxy = xmlrpclib.Server('http://pypi.python.org/pypi')
        self._downloads = {}

        self.first_upload = None
        self.first_upload_rel = None
        self.last_upload = None
        self.last_upload_rel = None

    @property
    def releases(self):
        """Retrieves the release number for each uploaded release"""

        result = self.proxy.package_releases(self.package_name, self.include_hidden)

        if len(result) == 0:
            # no matching package--search for possibles, and limit to 15 results
            results = self.proxy.search({
                'name': self.package_name,
                'description': self.package_name
            }, 'or')[:15]

            # make sure we only get unique package names
            matches = []
            for match in results:
                name = match['name']
                if name not in matches:
                    matches.append(name)

            # if only one package was found, return it
            if len(matches) == 1:
                self.package_name = matches[0]
                return self.releases

            error = """No such package found: %s

Possible matches include:
%s
""" % (self.package_name, '\n'.join('\t- %s' % n for n in matches))

            sys.exit(error)

        return result

    @property
    def downloads(self, force=False):
        """Calculate the total number of downloads for the package"""

        if len(self._downloads) == 0 or force:
            for release in self.releases:
                urls = self.proxy.release_urls(self.package_name, release)
                self._downloads[release] = 0
                for url in urls:
                    # upload times
                    uptime = datetime.strptime(url['upload_time'].value, "%Y%m%dT%H:%M:%S")
                    if self.first_upload is None or uptime < self.first_upload:
                        self.first_upload = uptime
                        self.first_upload_rel = release

                    if self.last_upload is None or uptime > self.last_upload:
                        self.last_upload = uptime
                        self.last_upload_rel = release

                    self._downloads[release] += url['downloads']

        return self._downloads

    def total(self):
        return sum(self.downloads.values())

    def average(self):
        return self.total() / len(self.downloads)

    def max(self):
        return max(self.downloads.values())

    def min(self):
        return min(self.downloads.values())

    def stats(self):
        """Prints a nicely formatted list of statistics about the package"""

        self.downloads  # explicitly call, so we have first/last upload data
        fmt = locale.nl_langinfo(locale.D_T_FMT)
        sep = lambda s: locale.format('%d', s, 3)
        val = lambda dt: dt and dt.strftime(fmt) or '--'

        params = (
            self.package_name,
            val(self.first_upload),
            self.first_upload_rel,
            val(self.last_upload),
            self.last_upload_rel,
            sep(len(self.releases)),
            sep(self.max()),
            sep(self.min()),
            sep(self.average()),
            sep(self.total()),
        )

        print """PyPI Package statistics for: %s

    First Upload: %40s (%s)
    Last Upload:  %40s (%s)
    Number of releases: %34s
    Most downloads:    %35s
    Fewest downloads:  %35s
    Average downloads: %35s
    Total downloads:   %35s
""" % params


def main():
    PyPIDownloadAggregator('pg_fabrep').stats()

if __name__ == '__main__':
    main()


pg_fabrep/example_fabfile.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
from fabric.api import env, task
from pg_fabrep.tasks import *


@task
def example_cluster():
    env.configured = True  # don't change this, used for testing purposes
    # postgres version, can be: "9.0" or "9.1"
    # default: "9.1"
    #env.postgres_version = "9.1"

    # name of your cluster - only numbers, letters and underscores
    # default: main
    env.cluster_name = "main"

    # postgres cluster port
    # default: 5432
    #env.cluster_port = 5432

    # the IP of your master server
    env.pgmaster_ip = ""

    # master ssh port, will be used by repmgr/rsync from the slave to replicate the master
    # default: 22
    #env.master_ssh_port = 22

    # master node number
    # default: 1
    #env.master_node_number = 1

    # user@host used by fabric to establish a ssh tunnel between the machine from where the "setup" is launched and the master server
    # default: "root@<env.pgmaster_ip>"
    #env.pgmaster_user_host = "root@%s" % env.pgmaster_ip

    # master pgconf path
    # default: "/etc/postgresql/<env.postgres_version>/<cluster_name>/"
    #env.master_pgconf_path = "/etc/postgresql/%s/%s/" % (env.postgres_version, env.cluster_name)

    # master pgdata path
    # default: "/var/lib/postgresql/<env.postgres_version>/<env.cluster_name>/"
    #env.master_pgdata_path = "/var/lib/postgresql/%s/%s/" % (env.postgres_version, env.cluster_name)

    # the IP of your slave server
    env.pgslave_ip = ""

    # slave node number
    # default: 2
    #env.slave_node_number = 2

    # user@host used by fabric to establish a ssh tunnel between the machine from where the "setup" is launched and the slave server
    # default: "root@<env.pgslave_ip>"
    #env.pgslave_user_host = "root@%s" % env.pgslave_ip

    # slave pgconf path
    # default: "<env.master_pgconf_path>"
    #env.slave_pgconf_path = env.master_pgconf_path

    # slave pgdata path
    # default: "<env.master_pgdata_path>"
    #env.slave_pgdata_path = env.master_pgdata_path

    # sync database used by repmgr, will be created if not exists
    # default: "syncdb"
    #env.sync_db = "syncdb"

    # sync postgres user used by repmgr, will be created if not exists
    # default: "syncuser"
    #env.sync_user = "syncuser"

    # sync postgres user password
    # default: "syncpass"
    #env.sync_pass = "syncpass"

    # repmgr_deb
    # default:"postgresql-repmgr-<env.postgres_version>_1.0.0.deb"
    #env.repmgr_deb = "postgresql-repmgr-%s_1.0.0.deb" % env.postgres_version

    # always ask user for confirmation when run any tasks
    # default: True
    #env.ask_confirmation = False

    # verbose, if True will print some info during the setup
    # default: False
    #env.verbose = True


pg_fabrep/tasks.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import re
from os.path import abspath, dirname, isfile
from datetime import datetime
from fabric.api import env, puts, abort, task, put, hide, cd
from fabric.operations import sudo, settings, run
from fabric.contrib import console
from fabric.contrib.files import upload_template

from fabric.colors import _wrap_with, green, red

green_bg = _wrap_with('42')
red_bg = _wrap_with('41')
pg_fabrep_path = dirname(abspath(__file__))


 #########################
## START pg_fabrep tasks ##
 #########################


@task
def setup():
    parameter_default_values()
    #  test configuration start
    if not test_configuration():
        if not console.confirm("Configuration test %s! Do you want to continue?" % red_bg('failed'), default=False):
            abort("Aborting at user request.")
    #  test configuration end
    if env.ask_confirmation:
        if not console.confirm("Are you sure you want to setup %s cluster?" % green(env.cluster_name.upper()), default=False):
            abort("Aborting at user request.")
    puts(green_bg('Start setup...'))
    start_time = datetime.now()

    # Start configuring master server
    with settings(host_string=env.pgmaster_user_host):
        print "%s configuring master server!" % green_bg("Start")
        _verify_sudo()
        _common_setup()
        with settings(hide('running', 'stdout', 'stderr', 'warnings'), warn_only=True):
            res = sudo("pg_createcluster --start %(postgres_version)s %(cluster_name)s -p %(cluster_port)s" % env)
        if 'already exists' in res:
            puts(green("Cluster '%(cluster_name)s' already exists, will not be changed." % env))
        if 'Error: port %s is already used' % env.cluster_port in res:
            puts(red(res))
        with cd(env.master_pgdata_path):
            with settings(hide('running', 'stdout', 'stderr', 'warnings'), warn_only=True):
                run('''sudo -u postgres psql -p %(cluster_port)s -c "CREATE USER %(sync_user)s SUPERUSER ENCRYPTED PASSWORD '%(sync_pass)s';"''' % env, shell=False)
                run('''sudo -u postgres createdb -p %(cluster_port)s --owner %(sync_user)s --template template0 --encoding=UTF8 --lc-ctype=en_US.UTF-8 --lc-collate=en_US.UTF-8 %(sync_db)s''' % env, shell=False)
                run('''sudo -u postgres psql -p %(cluster_port)s -c "GRANT CREATE, CONNECT ON DATABASE %(sync_db)s TO %(sync_user)s WITH GRANT OPTION;"''' % env, shell=False)
        postgres_conf_file = 'conf/postgres/postgresql.conf'
        if not isfile(postgres_conf_file):
            postgres_conf_file = '%s/%s' % (pg_fabrep_path, postgres_conf_file)
        upload_template(postgres_conf_file, env.master_pgconf_path,
                            context=env, backup=False)
        if env.verbose:
            puts("Uploaded postgresql.conf from %s to %s" % (green(abspath(postgres_conf_file)), green(env.master_pgconf_path)))
        # start preparing pg_hba.conf file
        replication_hba = '%sreplication_hba.conf' % env.master_pgconf_path
        pg_hba = '%spg_hba.conf' % env.master_pgconf_path
        pg_hba_bck = '%spg_hba.conf.bck' % env.master_pgconf_path
        pg_hba_file = 'conf/postgres/pg_hba.conf'
        if not isfile(pg_hba_file):
            pg_hba_file = '%s/%s' % (pg_fabrep_path, pg_hba_file)
        upload_template(pg_hba_file, replication_hba,
                        context=env, backup=False)
        sudo('mv %s %s' % (pg_hba, pg_hba_bck))
        sudo("sed -n -e '/START REPLICATION RULES/,/END REPLICATION RULES/!p' %s > %s" % (pg_hba_bck, pg_hba))
        sudo("cat %s >> %s" % (replication_hba, pg_hba))
        sudo("rm %s" % replication_hba)
        # upload repmgr.conf on master server
        repmgr_context = dict(cluster_name=env.cluster_name,
                              node_number=env.master_node_number,
                              sync_user=env.sync_user,
                              sync_db=env.sync_db,
                              sync_pass=env.sync_pass,
                              ssh_port=22,  # we don't need here
                              )
        repmgr_conf_file = 'conf/repmgr/repmgr.conf'
        if not isfile(repmgr_conf_file):
            repmgr_conf_file = '%s/%s' % (pg_fabrep_path, repmgr_conf_file)
        upload_template(repmgr_conf_file, env.master_pgdata_path,
                        context=repmgr_context, backup=False)
        sudo("""echo -e "# Added by pg_fabrep\nexport PGDATA='/var/lib/postgresql/%(postgres_version)s/%(cluster_name)s'\nexport PGPORT=%(cluster_port)s">/var/lib/postgresql/.bash_profile""" % env)
        sudo("pg_ctlcluster %(postgres_version)s %(cluster_name)s restart" % env)

    # Start configuring the slave
    with settings(host_string=env.pgslave_user_host):
        print "%s configuring slave server!" % green_bg("Start")
        _verify_sudo()
        _common_setup()
        sudo('rm -rf %s' % env.slave_pgdata_path)
        sudo('mkdir -p %s' % env.slave_pgdata_path)
        sudo('chown postgres:postgres %s' % env.slave_pgdata_path)
        _standby_clone()
        finish_configuring_slave()  # you can start it manually

    #  TODO: ...
    # # repmgr register master
    # with settings(host_string=env.master_host_local_ip):
    #     sudo('repmgr -f %srepmgr.conf --verbose master register' % \
    #          env.master_pgdata_path, user='postgres')
    # # repmgr register slave
    # with settings(host_string=env.slave_host):
    #     sudo('repmgr -f %srepmgr.conf --verbose standby register' % \
    #          env.slave_pgdata_path, user='postgres')
    end_time = datetime.now()
    finish_message = '[%s] Correctly finished in %i seconds' % \
    (green(end_time.strftime('%H:%M:%S')), (end_time - start_time).seconds)
    puts(finish_message)


@task
def test_configuration():
    errors = []
    if not env.get('configured'):
        puts(red('Configuration missing! Please read README.rst first or go ahead at your own risk.'))
        return False

    supported_postgres_versions = ("9.0", "9.1")
    if not env.postgres_version:
        errors.append('Postgres version missing')
    elif env.postgres_version not in supported_postgres_versions:
        errors.append('Wrong postgres version, supported versions are: "%s"' % '", "'.join(supported_postgres_versions))
    if not env.cluster_name:
        errors.append('Cluster name missing')
    elif not re.search(r"^\w+$", env.cluster_name):
        errors.append("%s is not a valid app name. Please use only numbers, letters and underscores." % red_bg(env.cluster_name))

    if not env.get('pgmaster_ip'):
        errors.append('Master server IP missing')
    if not env.get('master_ssh_port'):
        errors.append('Master server ssh port missing')
    if not env.get('master_node_number'):
        errors.append('Incorrect master node number')
    if not env.get('pgmaster_user_host'):
        errors.append('Incorrect master user@host')
    if not env.get('master_pgconf_path'):
        errors.append('Incorrect master pgconf path')
    if not env.get('master_pgdata_path'):
        errors.append('Incorrect master pgdata path')
    if not env.get('pgslave_ip'):
        errors.append('Slave server IP missing')
    if not env.get('slave_node_number'):
        errors.append('Incorrect slave node number')
    if not env.get('pgslave_user_host'):
        errors.append('Incorrect slave user@host')
    if not env.get('slave_pgconf_path'):
        errors.append('Incorrect slave pgconf path')
    if not env.get('slave_pgdata_path'):
        errors.append('Incorrect slave pgdata path')
    if not env.get('sync_db'):
        errors.append('Sync database name missing')
    if not env.get('sync_user'):
        errors.append('Sync user missing')
    if not env.get('sync_pass'):
        errors.append('Sync password missing')
    if not env.get('repmgr_deb'):
        errors.append('Repmgr deb file path missing')

    # print some feedback
    if errors:
        puts(red('Configuration test revealed %i errors:' % len(errors)))
        puts('%s\n\n* %s\n' % ('-' * 37, '\n* '.join(errors)))
        puts('-' * 40)
        puts('Please fix them or go ahead at your own risk.')
        return False
    puts(green('Configuration tests passed!'))
    print_configuration()
    return True


@task
def print_configuration():
    parameters_info = []
    parameters_info.append(("Postgres version", env.postgres_version))
    parameters_info.append(("Cluster name", env.cluster_name))
    parameters_info.append(("Cluster port", env.cluster_port))
    parameters_info.append(("--- Master Configuration", "----------"))
    parameters_info.append(("Master server IP", env.pgmaster_ip))
    parameters_info.append(("Master server ssh port", env.master_ssh_port))
    parameters_info.append(("Master node number", env.master_node_number))
    parameters_info.append(("Master user@host", env.pgmaster_user_host))
    parameters_info.append(("Master pgconf path", env.master_pgconf_path))
    parameters_info.append(("Master pgdata path", env.master_pgdata_path))
    parameters_info.append(("--- Slave Configuration", "----------"))
    parameters_info.append(("Slave server IP", env.pgslave_ip))
    parameters_info.append(("Slave node number", env.slave_node_number))
    parameters_info.append(("Slave user@host", env.pgslave_user_host))
    parameters_info.append(("Slave pgconf path", env.slave_pgconf_path))
    parameters_info.append(("Slave pgdata path", env.slave_pgdata_path))
    parameters_info.append(("--- Repmgr Configuration", "----------"))
    parameters_info.append(("Sync database name", env.sync_db))
    parameters_info.append(("Sync user", env.sync_user))
    parameters_info.append(("Sync password", env.sync_pass))
    parameters_info.append(("Repmgr deb file path", env.repmgr_deb))

    # print collected info
    puts('========== START YOUR CONFIGURATION ============')
    for parameter in parameters_info:
        parameter_formatting = "'%s'" if isinstance(parameter[1], str) else "%s"
        parameter_value = parameter_formatting % parameter[1]
        puts('%s %s' % (parameter[0].ljust(27), green(parameter_value)))
    puts('========== END YOUR CONFIGURATION ==============')


def parameter_default_values():
    if 'postgres_version' not in env:
        env.postgres_version = '9.1'
    if 'cluster_name' not in env:
        env.cluster_name = 'main'
    if 'cluster_port' not in env:
        env.cluster_port = 5432
    if 'pgmaster_ip' not in env:
        env.pgmaster_ip = ''
    if 'master_ssh_port' not in env:
        env.master_ssh_port = 22
    if 'master_node_number' not in env:
        env.master_node_number = 1
    if 'pgmaster_user_host' not in env:
        env.pgmaster_user_host = "root@%s" % env.pgmaster_ip
    if 'master_pgconf_path' not in env:
        env.master_pgconf_path = '/etc/postgresql/%s/%s/' % (env.postgres_version, env.cluster_name)
    if 'master_pgdata_path' not in env:
        env.master_pgdata_path = '/var/lib/postgresql/%s/%s/' % (env.postgres_version, env.cluster_name)
    if 'pgslave_ip' not in env:
        env.pgslave_ip = ''
    if 'slave_node_number' not in env:
        env.slave_node_number = 2
    if 'pgslave_user_host' not in env:
        env.pgslave_user_host = "root@%s" % env.pgslave_ip
    if 'slave_pgconf_path' not in env:
        env.slave_pgconf_path = env.master_pgconf_path
    if 'slave_pgdata_path' not in env:
        env.slave_pgdata_path = env.master_pgdata_path
    if 'sync_db' not in env:
        env.sync_db = 'syncdb'
    if 'sync_user' not in env:
        env.sync_user = 'syncuser'
    if 'sync_pass' not in env:
        env.sync_pass = 'syncpass'
    if 'repmgr_deb' not in env:
        env.repmgr_deb = "postgresql-repmgr-%s_1.0.0.deb" % env.postgres_version
    if 'ask_confirmation' not in env:
        env.ask_confirmation = True
    if 'verbose' not in env:
        env.verbose = False


@task
def finish_configuring_slave():
    parameter_default_values()
    with settings(host_string=env.pgslave_user_host):
        # upload repmgr.conf on slave server
        repmgr_context = dict(cluster_name=env.cluster_name,
                              node_number=env.slave_node_number,
                              sync_user=env.sync_user,
                              sync_db=env.sync_db,
                              sync_pass=env.sync_pass,
                              ssh_port=env.master_ssh_port,
                              )
        repmgr_conf_file = 'conf/repmgr/repmgr.conf'
        if not isfile(repmgr_conf_file):
            repmgr_conf_file = '%s/%s' % (pg_fabrep_path, repmgr_conf_file)
        upload_template(repmgr_conf_file, env.master_pgdata_path,
                        context=repmgr_context, backup=False)
        slave_postgresql_conf = "%spostgresql.conf" % env.slave_pgdata_path
        slave_postgresql_conf_bck = "%spostgresql.conf.bck" % env.slave_pgdata_path
        sudo('mv %s %s' % (slave_postgresql_conf, slave_postgresql_conf_bck))
        sudo("sed '/hot_standby =/c hot_standby = on' %s > %s" % \
             (slave_postgresql_conf_bck, slave_postgresql_conf))
        sudo("mkdir -p %s" % env.slave_pgconf_path)
        sudo("cp %spg_hba.conf %s" % (env.slave_pgdata_path, env.slave_pgconf_path))
        sudo("cp %spg_ident.conf %s" % (env.slave_pgdata_path, env.slave_pgconf_path))
        sudo("cp %spostgresql.conf %s" % (env.slave_pgdata_path, env.slave_pgconf_path))
        run("sudo -u postgres pg_ctl -D /var/lib/postgresql/%(postgres_version)s/%(cluster_name)s/ start" % env)


def _verify_sudo():
    ''' we just check if the user is sudoers '''
    sudo('cd .')


def _add_postgres9_ppa():
    ''' add postgresql 9.x ppa '''
    if 'Ubuntu 12.04' not in sudo('cat /etc/issue.net'):
        sudo('add-apt-repository ppa:pitti/postgresql')


def _install_dependencies():
    ''' Ensure those Debian/Ubuntu packages are installed '''
    packages = [
        'postgresql-%s' % env.postgres_version,
        'libpq-dev',
        'postgresql-server-dev-%s' % env.postgres_version,
        'postgresql-contrib-%s' % env.postgres_version,
        'libxslt-dev',
        'libxml2-dev',
        'libpam-dev',
        'libedit-dev',
    ]
    sudo("apt-get update")
    sudo("apt-get -y install %s" % " ".join(packages))
    #sudo("pip install --upgrade pip")


def _common_setup():
    sudo("apt-get -y install python-software-properties")
    _add_postgres9_ppa()
    _install_dependencies()
    # create a symlink in /usr/bin/ for /usr/lib/postgresql/<postgres_version>/bin/pg_ctl
    sudo('ln -sf /usr/lib/postgresql/%s/bin/pg_ctl /usr/bin/' % env.postgres_version)
    # install repmgr
    remote_repmgr_deb = '/tmp/%s' % env.repmgr_deb
    if isfile(env.repmgr_deb):
        local_repmgr_deb = env.repmgr_deb
    else:
        local_repmgr_deb = '%s/deb/%s' % (pg_fabrep_path, env.repmgr_deb)
    put(local_repmgr_deb, remote_repmgr_deb)
    run('dpkg -i --force-overwrite %s' % remote_repmgr_deb)
    sudo("rm %s" % remote_repmgr_deb)


def _standby_clone():
    """ With "node1" server running, we want to use the clone standby
    command in repmgr to copy over the entire PostgreSQL database cluster
    onto the "node2" server. """
    # manualy:
    # $ mkdir -p /var/lib/postgresql/9.1/testscluster/
    # $ rsync -avz --rsh='ssh -p2222' root@12.34.56.789:/var/lib/postgresql/9.1/testscluster/ /var/lib/postgresql/9.1/testscluster/

    with settings(hide('running', 'stdout', 'stderr', 'warnings'), warn_only=True):
        puts(green('Start cloning the master'))
        repmgr_clone_command = 'repmgr -D %(slave_pgdata_path)s -d %(sync_db)s -p %(cluster_port)s -U %(sync_user)s -R postgres --verbose standby clone %(pgmaster_ip)s' % env
        puts(green(repmgr_clone_command))
        puts("-" * 40)
        res = sudo(repmgr_clone_command, user='postgres')
        if 'Can not connect to the remote host' in res or 'Connection to database failed' in res:
            puts("-" * 40)
            puts(green(repmgr_clone_command))
            puts("-" * 40)
            puts("Master server is %s reachable." % red("NOT"))
            puts("%s you can try to CLONE the slave manually [%s]:" % (green("BUT"), red("at your own risk")))
            puts("On the slave server:")
            puts("$ sudo -u postgres rsync -avz --rsh='ssh -p%(master_ssh_port)s' postgres@%(pgmaster_ip)s:%(master_pgdata_path)s %(slave_pgdata_path)s --exclude=pg_xlog* --exclude=pg_control --exclude=*.pid" % env)
            puts("Here:")
            puts("$ fab <cluster_task_name> finish_configuring_slave")
            abort("STOP...")
